# AID-TF
This repository contains the codes for this paper:
"Integrative Data Analysis to Uncover Transcription Factors Involved in Gene Dysregulation of Nine Autoimmune and Inflammatory Diseases".<br>
In the following the description of folders and codes are provided:<br>
**Codes:**<br>
This directory contains all of the codes for the analysis and plotting the results. Each code is described in the following:<br>
**first_test.R**: This code performs the first step analysis which calculates the P values of enrichment of DHS samples for all samples in the dataset for a trait of interest. The inputs to the code are the trait name, the directory of DHS samples, the list of all DHS samples as a .txt file, address of the files having all immunochip data (reference) and Ci-SNPs (test) of all traits, and the output directory. This code calculates the P values of enrichment of all samples for the trait of interest and writes lists containing the list of processed samples, P values of enrichment of samples for the trait of interest, list of enriched samples, number of test SNPs (Ci-SNPs) and reference SNPs (all immunochip SNPs) matching the effect SNPs of samples. <br>
**second_test.R**: This code calculates the enrichment of motifs in the enriched samples for a trait of interest. The inputs to this code are trait name, the directory of DHS samples, the list of all DHS samples as a .txt file, address of the files having all immunochip data (reference) and Ci-SNPs (test) of all traits, list of enriched samples for the trait, and an output directory. This code calculates the enrichment of motifs within each enriched samples, creates a subdirectory within the output directory for each samples, and writes the list of motifs and their p values of enrichment for the sample in the subdirectory. <br>
**final_complete_dataframe_prep.R**: This code integrates the results of first and second step analysis an creates a table having all results for each trait comprising full list of significant AIDsâ€™ risk variants and their genomic locations, affected motifs and transcription factors (genomic position, motif id, and transcription factor name), relevant samples and cell types, P values of relevance of motifs and samples, and the trait of interest. The inputs of this code are trait name, the directory of DHS samples, the list of all DHS samples as a .txt file, address of the files having all immunochip data (reference) and Ci-SNPs (test) of all traits, the file having the P values of enrichment of samples for the trait, the files comprising list of motifs and their P values of enrichment within samples, and an output directory for the trait of interest. The code writes the full table and a table used for great pathway analysis in next steps for each trait in the given output directory. The full tables of traits can be merged into Supplementary Table S1.<br>
s
s
