# AID-TF
This repository contains the codes for this paper:
"Integrative Data Analysis to Uncover Transcription Factors Involved in Gene Dysregulation of Nine Autoimmune and Inflammatory Diseases".<br>
In the following the description of folders and codes are provided:<br>
**Codes:**<br>
This directory contains all of the codes for the analysis and plotting the results. Each code is described in the following:<br>
**first_test.R**: This code performs the first step analysis which calculates the P values of enrichment of DHS samples for all samples in the dataset for a trait of interest. The inputs to the code are the trait name, the directory of DHS samples, the list of all DHS samples as a .txt file, address of the files having all immunochip data (reference) and Ci-SNPs (test) of all traits, and the output directory. This code calculates the P values of enrichment of all samples for the trait of interest and writes lists containing the list of processed samples, P values of enrichment of samples for the trait of interest, list of enriched samples, number of test SNPs (Ci-SNPs) and reference SNPs (all immunochip SNPs) matching the effect SNPs of samples. <br>
**second_test.R**: This code calculates the enrichment of motifs in the enriched samples for a trait of interest. The inputs to this code are trait name, the directory of DHS samples, the list of all DHS samples as a .txt file, address of the files having all immunochip data (reference) and Ci-SNPs (test) of all traits, list of enriched samples for the trait, and an output directory. This code calculates the enrichment of motifs within each enriched samples, creates a subdirectory within the output directory for each samples, and writes the list of motifs and their p values of enrichment for the sample in the subdirectory. <br>
**final_complete_dataframe_prep.R**: This code integrates the results of first and second step analysis and creates a table having all of the results for each trait comprising significant AIDsâ€™ risk variants and their genomic locations, affected motifs and transcription factors (genomic position, motif id, and transcription factor name), relevant samples and cell types, P values of relevance of motifs and samples, and the trait of interest. The inputs of this code are trait name, the directory of DHS samples, the list of all DHS samples as a .txt file, address of the files having all immunochip data (reference) and Ci-SNPs (test) of all traits, the file having the P values of enrichment of samples for the trait, the files comprising list of motifs and their P values of enrichment within samples, and an output directory for the trait of interest. The code writes the full table and a table used for great pathway analysis in next steps for each trait in the given output directory. The full tables of traits can be merged into Supplementary Table S1.<br>
**first_test_allplots.R**: This code generates all of the figures for the first step results comprising P values of enrichment of samples for a trait. The inputs to this code are files having list of samples, and their P values of enrichment for traits, and output directory for figures. This code generates figures 1, supplementary figures s1 and s2, and cell enrichment tables. <br>
**second_test_allplots.R**: This code generates all of the figures for the second step analysis comprising P values of enrichment of motifs within enriched samples for a trait. The inputs of this code are the final data table comprising all results for all traits, list of all motifs, list of all samples, and trait-specific full results tables, and output directory. This code generates figures 2, 3, and supplementary figure s3. <br>
**traitvstrait.R**: This code compares the traits in terms of their enriched samples, cell types, and motifs and generates heatmaps for these comparisons. The inputs of this code are the final data table having all the results, list of traits, and output directories for generated figures. This code generates heatmaps of figure 4. <br>
**chromhmmplot.R**: This code generates the heatmaps showing the results of ChromHMM analysis. The inputs are files of the list of motifs and their TF names, directory of ChromHMM results of each trait, list of traits, and output directory. This code generates heatmaps showing the results of affected TF functions for each trait and stores them in the output directory. This code generates figure 5. <br>
**greatanalysis.R**: This code runs the great pathway analysis on affected DHS regions of traits and generates the results. The inputs are list of traits, full table results of traits, and output directory. This code generates figure 6 in the output directory.<br>
**Pathway_gene_tf.R**: This code identifies the genes involved in the affected samples of a trait, the TFs affecting those genes, and the distance between those genes and affected DHS sites. These results are generated for all traits. The inputs are the directories of trait-affected DHS sites (in .bed format), list of traits, final table of results, directories of trait-related pathways, TF-gene pair results directories, and output directory. This code generates and save tables 1 and 2 in the output directory. 
